### PR TITLE
chore: change the project structure

### PR DESCRIPTION
**Pull request by musale**
### Description of changes
- Move all source files/directories into the `src` folder.
- Move all test files/directories into the `test` folder.
- Update a few tests.
- Change the test framework from `mocha` and `chai` for assertions to https://vitest.dev.
- Set up `husky` to install git pre-commit hooks. Run code style check and linting in the pre-commit hook.
- Setup `eslint`, `prettier` and `vitest` with basic config and extend msgraph defaults.
- Update pipelines to use the new commands in the workflows.
- Change the `updateVersion` script from `powershell` to `nodejs`.
- Rename the source files to PascalCase.

### CODE REVIEW

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for starting this.
~~I believe we need to update the package.json for the main/types fields etc...~~
Maybe the npmignore as well
~~And why are we deleting the lock?~~

**Comment by musale** [comment]
> Thanks for starting this. I believe we need to update the package.json for the main/types fields etc... Maybe the npmignore as well And why are we deleting the lock?

I'll update all those as well. Normally I use `yarn` for local development. I assumed that this project is using `npm` cos of the `package-lock.json`. The deleted lock file is `yarn.lock`. However, I propose using `yarn` because of it's many modern features like zero installs for development.

**Comment by baywet** [comment]
@musale if we switch to yarn, I'll request that we do so across all repositories (kiota-typescript, this one, the service libs, etc...)
The challenge I previously had was that ONLY kiota-typescript had been setup with yarn and that was inconsistent across repos (and annoying during local dev loops). Let's treat this change separately for now please.
I have stricken my comments about the package fields since it was already done, as well as the one about the lock file, sorry about the confusion.
I think your TODOs and the other comments I've made are still applicable.

File: .github/workflows/build_test_validate.yml [hunk 1]
```diff
@@ -18,14 +18,9 @@ jobs:
           uses: actions/setup-node@v4
           with:
             node-version: ${{ matrix.node-version }}
-        - run: .\scripts\updateVersion.ps1
-          shell: pwsh
-          working-directory: ./
         - run: npm ci
-        - run: ./scripts/updateVersion.ps1
-          shell: pwsh
         - run: npm run build
```
Comment (baywet) [inline .github/workflows/build_test_validate.yml:22]: shouldn't we have a step to run the update version script (JS) here?

File: package.json [hunk 2]
```diff
@@ -5,13 +5,35 @@
...
+    "dist",
+    "src"
+  ],
   "scripts": {
-    "build": "tsc -p tsconfig.json",
-    "test": "mocha tests/**/*.js",
-    "lint": "eslint . --ext .ts",
-    "lint:fix": "eslint . --ext .ts --fix"
+    "prepare": "husky",
+    "build": "npm run updateVersion && npm run prettier:check && npm run clean && npm run build:compile",
```
Comment (musale) [inline package.json:20]: @baywet the build step runs the script to update the version first before "building" commands.

**Comment by musale** [comment]
> @musale if we switch to yarn, I'll request that we do so across all repositories (kiota-typescript, this one, the service libs, etc...) The challenge I previously had was that ONLY kiota-typescript had been setup with yarn and that was inconsistent across repos (and annoying during local dev loops). Let's treat this change separately for now please. I have stricken my comments about the package fields since it was already done, as well as the one about the lock file, sorry about the confusion. I think your TODOs and the other comments I've made are still applicable.

In this case, then let's leave it. I did change the test framework from mocha and chai to https://vitest.dev. There weren't any tests but from the other js repo, we would need to test async and some browser code in here, so I picked a framework that supports all of that out of the box with minimal setup for es module project. I also added husky which installs git hooks that locally run prettier and lint commands before you committing.

**Review by baywet — DISMISSED** [review]
Thanks for all the great changes here! My TypeScript knowledge is a bit rusty, and that ecosystem keeps changing so fast...

**Review by baywet — APPROVED** [review]

**Pull request merged**